### ISSUE TITLE
buildah bud 'RUN' incorrectly executing entrypoint

### ISSUE THREAD

**Issue by pixdrift** [projectatomic/buildah#607]
**Description**
When an entry point is configured in a Dockerfile executed by `buildah bud`, the `RUN` commands in the Dockerfile triggers execution of the configured `ENTRYPOINT` value.

Credit to @adelton for initial report and reproducer.

**Steps to reproduce the issue:**
1. `buildah bud` the following Docker file
```
FROM centos:7
RUN uuidgen > /build-id-pre-entrypoint
ENTRYPOINT [ "/non/existent/entrypoint" ]
RUN uuidgen > /build-id-post-entrypoint
```

Result:
```
STEP 1: FROM centos:7
STEP 2: RUN uuidgen > /build-id-pre-entrypoint
STEP 3: ENTRYPOINT [ "/non/existent/entrypoint" ]
STEP 4: RUN uuidgen > /build-id-post-entrypoint
container_linux.go:348: starting container process caused "exec: \"/non/existent/entrypoint\": stat /non/existent/entrypoint: no such file or directory"
error building at step {Env:[PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin] Command:run Args:[uuidgen > /build-id-post-entrypoint] Flags:[] Attrs:map[] Message:RUN uuidgen > /build-id-post-entrypoint Original:RUN uuidgen > /build-id-post-entrypoint}: exit status 1
```

**Describe the results you received:**
Second `RUN` command in Dockerfile executes using `ENTRYPOINT` configured in previous step

`RUN` command in Dockerfile should not execute using `ENTRYPOINT` when built with `buildah bud`.

**Describe the results you expected:**
Successful `RUN` execution without use of configured `ENTRYPOINT` to match Docker behaviour.

**Output of `rpm -q buildah` or `apt list buildah`:**

```
buildah-0.16-201804191042.gite008b73.el7.x86_64
```

**Output of `buildah version`:**

```
Version:       0.16
Go Version:    go1.8.3
Image Spec:    1.0.0
Runtime Spec:  1.0.0
Git Commit:    e008b73
Built:         Thu Apr 19 21:58:48 2018
OS/Arch:       linux/amd64
```

**Output of `cat /etc/*release`:**

```
CentOS Linux release 7.4.1708 (Core) 
```

**Comment by pixdrift** [comment]
Execution of same Dockerfile with `docker build`:
```
# docker build budrun
Sending build context to Docker daemon 2.048 kB
Step 1/4 : FROM centos:7
 ---> e934aafc2206
Step 2/4 : RUN uuidgen > /build-id-pre-entrypoint
 ---> Running in 29118e9d1b6a

 ---> 15d4206ea973
Removing intermediate container 29118e9d1b6a
Step 3/4 : ENTRYPOINT /non/existent/entrypoint
 ---> Running in 2b595e7f6b67
 ---> bd7259a96e4c
Removing intermediate container 2b595e7f6b67
Step 4/4 : RUN uuidgen > /build-id-post-entrypoint
 ---> Running in 3b76b323cee4

 ---> ffdc67989d22
Removing intermediate container 3b76b323cee4
Successfully built ffdc67989d22
```

**Comment by pixdrift** [comment]
This looks related to entrypoint change that fixed 'buildah run' behaviour.
https://github.com/projectatomic/buildah/commit/15e1054820a0442ce179ca44ddd870b8ff56fa8c
  
Code block here:
https://github.com/projectatomic/buildah/blob/15e1054820a0442ce179ca44ddd870b8ff56fa8c/run.go#L331-L343
  
Looks like `buildah bud` calls the same run fuction
https://github.com/projectatomic/buildah/blob/83d7d100df1c048d34b6efd6ac7b8d0305ad03f1/imagebuildah/build.go#L422

Interested in advice @rhatdan @TomSweeneyRedHat on how this should be approached. It needs to be determined how the `run` was called (by `buildah bud` or `buildah run`) and the behaviour of the entrypoint append adjusted accordingly.

To summarise the desired behaviour as I understand it:
`RUN` commands defined in Dockerfiles should never use entrypoint
`buildah run` execution uses entrypoint if it is configured

**Comment by rhatdan** [comment]
I believe buldah run and RUN command should use the exact same behaviour.  buildah run is just supposed to be an BASH version of the Dockerfile RUN.  @nalind do you agree.  

I agree that we need to match Docker functionality, though.

**Comment by pixdrift** [comment]
@rhatdan, the modification made was to make `buildah run` match `docker run` by passing parameters to the entrypoint. I also would expected `RUN` and `${tool} run` to behave the same, and was surprised by this issue.

**Comment by rhatdan** [comment]
Well podman run should match docker run, but buildah run should match Dockerfile RUN.

**Comment by pixdrift** [comment]
@rhatdan, if that's the case I can raise a PR to back out the change that was implemented to resolve issue https://github.com/projectatomic/buildah/issues/564

**Comment by pixdrift** [comment]
Can confirm that revert of 97a0790def2935571469f9cc3d5018a61cfd390b returns behaviour to `buildah bud RUN = buildah run`

Reproducer confirmed fixed after reverted commit:
```
STEP 1: FROM centos:7
STEP 2: RUN uuidgen > /build-id-pre-entrypoint
STEP 3: ENTRYPOINT [ "/non/existent/entrypoint" ]
STEP 4: RUN uuidgen > /build-id-post-entrypoint
STEP 5: COMMIT containers-storage:[overlay@/var/lib/buildah/storage+/var/run/buildah/storage:overlay.override_kernel_check=true]@7aa1c7373614ce494ae927fa537ca39907af442c237351d698f342400d5f16a3
Getting image source signatures
Skipping fetch of repeat blob sha256:43e653f84b79ba52711b0f726ff5a7fd1162ae9df4be76ca1de8370b8bbf9bb0
Copying blob sha256:2e5b477c32ec15a7fa755b6538d94e4d13396734b95c503a5b22618b7e059678
 270 B / 270 B [============================================================] 0s
Copying config sha256:e467b3f2801c661a5d30f4519ef7ca0bc3c8dc1c999da84ab10639101e323248
 1.15 KiB / 1.15 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
7aa1c7373614ce494ae927fa537ca39907af442c237351d698f342400d5f16a3
```

**Comment by pixdrift** [comment]
@adelton, this should be resolved. Can you re-test?

**Comment by rhatdan** [comment]
Closing issue.  @adelton Please reopen if you disagree.

**Comment by pixdrift** [comment]
@rhatdan @TomSweeneyRedHat, can this be re-opened? there is a lingering issue here.

I had a small niggling issue I couldn't narrow down and I have finally created a reproducer for it. This works off the following assumption of desired behaviour mentioned above.

```
buildah run = Dockerfile RUN
podman run = docker run
```

The following Docker file (intentionally broken syntax on last line):
```
FROM centos:7
RUN uuidgen > /build-id-pre-entrypoint
ENTRYPOINT [ "/non/existent/entrypoint" ]
RUN
```

**buildah bud result:**
```
STEP 1: FROM centos:7
STEP 2: RUN uuidgen > /build-id-pre-entrypoint
STEP 3: ENTRYPOINT [ "/non/existent/entrypoint" ]
STEP 4: RUN
/bin/sh: -c: option requires an argument
error building at step {Env:[PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin] Command:run Args:[] Flags:[] Attrs:map[] Message:RUN  Original:RUN}: exit status 2
```
This `Dockerfile RUN` attempts to run an empty value, and _doesn't_ insert the `entrypoint` if `run` parameter isn't provided, resulting in an error.

The following is a buildah shell file to replicate the Dockerfile above:
```
#!/bin/bash
set -ex
container=$(buildah from centos:7)
buildah run ${container} uuidgen > /build-id-pre-entrypoint
buildah config --entrypoint "/non/existent/entrypoint" ${container}
buildah run ${container}
```

**Execution result:**
```
++ buildah from centos:7
+ container=centos-working-container-11
+ buildah run centos-working-container-11 uuidgen
+ buildah config --entrypoint /non/existent/entrypoint centos-working-container-11
+ buildah run centos-working-container-11
/bin/sh: /non/existent/entrypoint: No such file or directory
```

Rather than match the behaviour of Dockerfile RUN, it attempts to execute `entrypoint` because the command parameter is empty, it should instead error, consistent with Dockerfile RUN.

**Comment by rhatdan** [comment]
Yes I agree.  Another way to look at this is we should make `buildah run CTR` always require a command.

**Comment by pixdrift** [comment]
Agreed. Looking at the code `cmd` may also behave the same way, expanding the tests to check this also.

**Comment by pixdrift** [comment]
As suspected, `cmd` suffers the same affliction.
```
STEP 1: FROM centos:7
STEP 2: RUN uuidgen > /build-id-pre-entrypoint
STEP 3: CMD [ "/non/existent/command" ]
STEP 4: RUN
/bin/sh: -c: option requires an argument
```

```
++ buildah from centos:7
+ container=centos-working-container-12
+ buildah run centos-working-container-12 uuidgen
+ buildah config --cmd /non/existent/command centos-working-container-12
+ buildah run centos-working-container-12
container_linux.go:348: starting container process caused "exec: \"/non/existent/command\": stat /non/existent/command: no such file or directory"
```

**Comment by kbaegis** [comment]
One thing I've been noticing is that ENV may not be inherited from the parent image. IDK if that helps, but please let me know if you're noticing this as well.

**Comment by pixdrift** [comment]
@kbaegis, thanks for the feedback, can you give a specific example?

Is the issue you are seeing that the `RUN` command in a Dockerfile isn't using the `ENV` configured in the source image? Do you have a reproducer/test? and can you confirm that the behaviour differs from `docker build `(if it does?)

**Comment by TomSweeneyRedHat** [comment]
@kbaegis +1 to @pixdrift's comments.  Saw you IRC after you'd left the channel this weekend.  That should work, but I'm not sure that I've tried it.  If you've a concrete example, that would be very helfpul.  Please do open an issue if you feel it's warranted, it sounds like it is.

**Comment by pixdrift** [comment]
@kbaegis, I setup a test for `ENV` configured in both the same Dockerfile and an image sourced using `FROM` and in both cases ENV behaved correctly, and the behaviour in `buildah bud` matched that of `docker build`.

If you can provide a specific reproducer I can take a closer look. I am going to add this `ENV` scenario to the `bud` CI tests.

**Comment by pixdrift** [comment]
This was resolved by #656 and new tests added to CI.
